### PR TITLE
Internal services

### PR DESCRIPTION
**Pull request by mawolf**
Adding support for internal services

### CODE REVIEW

File: jolie/src/jolie/Interpreter.java [hunk 1]
```diff
@@ -1182,26 +1213,33 @@ private boolean buildOOIT()
 				} else {
 					throw new InterpreterException( "Input compiled program is not a JOLIE program" );
 				}
-			} else {
+			}
+            else if (this.internalServiceProgram != null) {
+                program = this.internalServiceProgram;
+                OLParseTreeOptimizer optimizer = new OLParseTreeOptimizer( program );
```
Comment (fmontesi) [inline jolie/src/jolie/Interpreter.java]: Looks like this code is duplicated with the one in the next branch.

File: jolie/src/jolie/runtime/embedding/EmbeddedServiceLoader.java [hunk 2]
```diff
@@ -100,4 +107,77 @@ protected void setChannel( CommChannel channel )
...
+    public static abstract class EmbeddedServiceConfiguration
+    {
+        private final Constants.EmbeddedServiceType type;
+        
+        public EmbeddedServiceConfiguration(Constants.EmbeddedServiceType type)
+        {
+            this.type = type;
+        }
+        
+        public Constants.EmbeddedServiceType getType()
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/embedding/EmbeddedServiceLoader.java]: Please rename getType() to type()

File: jolie/src/jolie/runtime/embedding/EmbeddedServiceLoader.java [hunk 3]
```diff
@@ -100,4 +107,77 @@ protected void setChannel( CommChannel channel )
...
+         *
+         * @param type Type of embedded service, cannot be INTERNAL
+         * @param servicePath path of service
+         */
+        public ExternalEmbeddedServiceConfiguration( Constants.EmbeddedServiceType type, String servicePath )
+        {
+            super(type);
+            this.servicePath = servicePath;
+            
+            assert type != Constants.EmbeddedServiceType.INTERNAL;
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/embedding/EmbeddedServiceLoader.java]: Should probably throw IllegalArgumentException in case type is INTERNAL

File: libjolie/src/jolie/lang/parse/OLParser.java [hunk 4]
```diff
@@ -811,8 +812,160 @@ private void parseInterfaceOrPort()
...
+        InputPortInfo iport = null;
+        try {
+            iport = new InputPortInfo(
+                getContext(),
+                serviceName + "InputPort", //input port name
+                new URI( Constants.LOCAL_LOCATION_KEYWORD ),
+                null,
+                protocolConfiguration,
+                new InputPortInfo.AggregationItemInfo[] {},
+                new HashMap<String, String>() );
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/OLParser.java]: Use Collections.emptyMap()

File: libjolie/src/jolie/lang/parse/OLParser.java [hunk 5]
```diff
@@ -811,8 +812,160 @@ private void parseInterfaceOrPort()
...
+        String serviceName = token.content();
+        
+        //validate token
+        getToken();
+        eat(Scanner.TokenType.LCURLY, "{ expected");
+        
+        //initialize internal interface and interface list
+        List< InterfaceDefinition > interfaceList = new ArrayList< InterfaceDefinition >();
+        
+        OLSyntaxNode internal_main = null;
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/OLParser.java]: internal_main -> internalMain

File: libjolie/src/jolie/lang/parse/OLParser.java [hunk 6]
```diff
@@ -811,8 +812,160 @@ private void parseInterfaceOrPort()
...
+        SequenceStatement internal_init = null;
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/OLParser.java]: internal_init -> internalInit

Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/OLParser.java]: internal_init could probably just be OLSyntaxNode, SequenceStatement seems redundant.

File: libjolie/src/jolie/lang/parse/util/VisualizeAST.java [hunk 7]
```diff
@@ -0,0 +1,458 @@
+/***************************************************************************
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/util/VisualizeAST.java]: Where is this class used? Please remove it if unused (we could move it to an external library though maybe... but let's leave it for another patch).

File: libjolie/src/jolie/lang/parse/OLParser.java [hunk 8]
```diff
@@ -811,8 +812,160 @@ private void parseInterfaceOrPort()
...
+            throwException( "You must specify a main for internal service " + serviceName );
+      
+        Program internalServiceProgram = new Program(getContext());
+        
+        // copy children of parent to embedded service
+        for ( OLSyntaxNode child : program.children() ) {
+            if ( child instanceof TypeInlineDefinition ||
+                child instanceof InterfaceDefinition ||
+                child instanceof OutputPortInfo ||
+                child instanceof EmbeddedServiceNode )
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/OLParser.java]: This does not look like the right way of implementing the first requirement in Semantics of #27 .
It's going to duplicate embedded services, which could be a problem in case of sub-services written in Jolie (and even Java, although many are harmless, like Console).

What we should do is implement some mechanism for the internal service to wait for the bindings for the output ports at runtime. This doesn't look too easy, we need to think of the lifecycle defined in:

https://github.com/jolie/jolie/blob/master/jolie/src/jolie/process/InitDefinitionProcess.java

Embedded services are loaded before the init procedure starts, so the internal service can't wait for the parent's init to end before "copying over" the bindings for the output ports or we'll create a deadlock.

**Comment by fmontesi** [comment]
Great patch, thanks. :-)

Indentation looks weird sometimes. Please follow the style of the rest of the code (spaces in parentheses) and use tabs instead of spaces.

@jolie/developers More eyeballs on the patch are welcome.

**Comment by fmontesi** [comment]
The formatting rules I used (updated to last year I think, but netbeans should not have changed so much hopefully?..)

http://filebin.ca/22Ay89ym7QDR/formatting.zip

**Comment by mawolf** [comment]
Again, thank you for the comments and recommendations!

The latest commits should fix the concerns that you had. Let me know if further changes are needed!
